### ISSUE TITLE
Version pinning misses capability for Regular Expressions

### ISSUE THREAD

**Issue by megies** [conda/conda#2716]
I would like to restrict a package's version to release versions through pinning, excluding pre-release builds.
I think the only way to achieve this is using a regex for the version specification in the `pinned` file. But since there is no way to explicitly force a line in the pinned file to be marked a regex, [we can only get a regex-matching `VersionSpec` by satisfying this line in it's constructor](https://github.com/conda/conda/blob/231e9db898b3d7720b49e9e2050a88be6978fd38/conda/version.py#L319):
```
    elif '*' in spec.rstrip('*'):
```
So we would have to put multiple stars in the version spec but make sure that the resulting regex is still what we want to match.. plus, there's all kind of implicit replacementsdone to our regex, so this is not feasible I think.

Consider this example:
```
((maintenance_1.0.x)) wintermute:~$ conda search basemap
Fetching package metadata ...........
basemap                      1.0.6                np17py27_0  defaults        
                          .  1.0.7               np110py27_0  defaults        
                             1.0.8.dev           np110py27_0  conda-forge/linux-64 
                          *  1.0.8.dev0          np110py27_1  conda-forge/linux-64 
```
Let's assume I want to pin to any release version build of `1.0` series, excluding the "*.dev*" pre-release packages. I don't see a way to do that..

How about something like..
```
basemap ^1\.0\.[0-9]+$
```
and an if/else clause in the (admittedly already pretty cluttered) constructor of `VersionSpec`:
```python
elif spec.startswith("^") and spec.endswith("$"):
    self.spec = spec
    self.regex = re.compile(spec)
    self.match = self.regex_match_
```
That would enable people to use Regex in `pinned` file (detected by expicit start/end of string markers)..

**Comment by mcg1969** [comment]
Yeah, we never really intended to expose full regex matching for versions---just glob-style wildcards. I certainly wouldn't mind the notion of having some sort of explicit regex requests.

**Comment by megies** [comment]
I've pushed my proposal to a branch: https://github.com/megies/conda/tree/versionspec_regex

**Comment by kalefranz** [comment]
Actually, why don't you go ahead and make this a PR @megies.  I don't see any issues with the concept right now.

**Comment by megies** [comment]
I don't have push rights, so I can't change this issue into a PR. I'd rather avoid a duplicated ticket, would you mind to make that change for me?

Should've made it a PR right away I guess, but wanted to hear opinions first.

**Comment by mcg1969** [comment]
You can PR from a fork, no need for push rights

**Comment by megies** [comment]
> You can PR from a fork, no need for push rights

Yes, I know. Sorry for being unclear.. I meant transforming this very issue into a PR, which requires push rights to `conda/conda` ([e.g. with curl](http://stackoverflow.com/questions/4528869/how-do-you-attach-a-new-pull-request-to-an-existing-issue-on-github/4529172#4529172)):
```bash
$ curl --user "my_github_username" --request POST --data '{"issue": "2716", "head": "megies:versionspec_regex", "base": "master"}' https://api.github.com/repos/conda/conda/pulls
```

Sorry, should've just started it as a PR right away.. but wasn't sure if the change would be welcomed.

**Comment by mcg1969** [comment]
Oh, I see. Well, I never convert issues to PRs that way, I didn't even know it could be done! Learn something every day! I just create a PR and link to the corresponding issue in the PR text.

**Comment by mcg1969** [comment]
Regarding specifically the use of `^` and `$`: do we want to require both? What if we use an *or* instead of an *and* here? That way, if someone just wants to select a specific GIT tag, the could do; e.g., `-gdeadbee$`.

**Comment by megies** [comment]
> That way, if someone just wants to select a specific GIT commit, the could do, say, `-gdeadbee$`.

You could still do that with `^.*-gdeadbee$`. My line of thought was to make double-sure to avoid catching anything that might not be a regex deliberately invoked by the user.

**Comment by mcg1969** [comment]
OK, serves me right for not reading the full thread before I commented. I'm fine leaving them both in, but I made some comments on your branch.

**Comment by kalefranz** [comment]
> do we want to require both?

I like both.

**Comment by kalefranz** [comment]
@megies I probably said convert.  I meant create a new pull request, and then reference this issue.  That's totally fine.

**Comment by kalefranz** [comment]
Closed with #2933
